### PR TITLE
fix: Remove triple parsing of diagrams

### PR DESCRIPTION
**Pull request by sidharthv96**
## :bookmark_tabs: Summary

Many diagrams were parsed 2-3 times for each render. Removed that.

## :straight_ruler: Design Decisions

Moved the call to `init()` after `clear()`.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [ ] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/community/development.md#3-update-documentation) is used for all new features.
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

**Comment by sidharthv96** [comment]
Time taken to render diagrams in `demos/flowchart.html`

| Browser  | Before             | After              | Decrease |
| -------- | ------------------ | ------------------ | -------- |
| Chromium | 2289.7900001049043 | 1945.0699999332428 | 15%      |
| Firefox  | 1157.3             | 971.8              | 16%      |
| Webkit   | 2626.2             | 2525.5             | 3.8%     |

 <table>
<tr>
<td><details>
  <summary>Before</summary>
<h2>Chromium</h2> <br/>
render: 2157.5 ms <br/>
render: 2163.2000002861023 ms <br/>
render: 2151.5 ms <br/>
render: 2160.699999809265 ms <br/>
render: 2709.1000003814697 ms <br/>
render: 2367.5 ms <br/>
render: 2637.7000002861023 ms <br/>
render: 2195.1000003814697 ms <br/>
render: 2227.4000000953674 ms <br/>
render: 2128.199999809265 ms <br/>
Average: 2289.7900001049043 <br/>
 <hr/>
<h2>Firefox</h2> <br/>
render: 1245 ms <br/>
render: 1133 ms <br/>
render: 1161 ms <br/>
render: 1150 ms <br/>
render: 1167 ms <br/>
render: 1137 ms <br/>
render: 1151 ms <br/>
render: 1154 ms <br/>
render: 1125 ms <br/>
render: 1150 ms <br/>
Average: 1157.3 <br/>
 <hr/>
<h2>Webkit</h2> <br/>
render: 2758 ms <br/>
render: 2637 ms <br/>
render: 2653 ms <br/>
render: 2614 ms <br/>
render: 2602 ms <br/>
render: 2549 ms <br/>
render: 2588 ms <br/>
render: 2610 ms <br/>
render: 2595.0000000000005 ms <br/>
render: 2656 ms <br/>
Average: 2626.2 <br/>


</details> </td>
<td><details>
  <summary>After</summary>
<h2>Chromium</h2> <br/>
render: 1963.9000000953674 ms <br/>
render: 1891.0999999046326 ms <br/>
render: 1998 ms <br/>
render: 1934.7999997138977 ms <br/>
render: 1982.5 ms <br/>
render: 1944.7999997138977 ms <br/>
render: 1950.5 ms <br/>
render: 1915.5 ms <br/>
render: 1911.5 ms <br/>
render: 1958.0999999046326 ms <br/>
Average: 1945.0699999332428 <br/>
 <hr/>
<h2>Firefox</h2> <br/>
render: 1037 ms <br/>
render: 954 ms <br/>
render: 1026 ms <br/>
render: 943 ms <br/>
render: 954 ms <br/>
render: 939 ms <br/>
render: 973 ms <br/>
render: 986 ms <br/>
render: 964 ms <br/>
render: 942 ms <br/>
Average: 971.8 <br/>
 <hr/>
<h2>Webkit</h2> <br/>
render: 2535 ms <br/>
render: 2583 ms <br/>
render: 2557 ms <br/>
render: 2536 ms <br/>
render: 2540 ms <br/>
render: 2554 ms <br/>
render: 2512 ms <br/>
render: 2479 ms <br/>
render: 2486 ms <br/>
render: 2473 ms <br/>
Average: 2525.5 <br/>
</details>
</td>
</tr>
</table>

**Comment by sidharthv96** [comment]
Related
- https://github.com/mermaid-js/mermaid/pull/4587

**Comment by Yokozuna59** [comment]
Awesome!

I saw some diagrams have `init` function in it's definition, and it's clearing the DB there, is it related to this PR? i.e., [flowDiagram-v2](https://github.com/mermaid-js/mermaid/blob/f8ebfeefec466d5001939f5a520a89cee2e4639d/packages/mermaid/src/diagrams/flowchart/flowDiagram-v2.ts#L22).

**Comment by sidharthv96** [comment]
> Awesome!
> 
> I saw some diagrams have `init` function in it's definition, and it's clearing the DB there, is it related to this PR? i.e., [flowDiagram-v2](https://github.com/mermaid-js/mermaid/blob/f8ebfeefec466d5001939f5a520a89cee2e4639d/packages/mermaid/src/diagrams/flowchart/flowDiagram-v2.ts#L22).

Yeah, I noticed that, but didn't want to change too much in this PR. That can be removed and tested in an upcoming flowDB cleanup PR (that I'm working on).

**Comment by nirname** [comment]
Awesome. What I previously did was only initial cleanup of extra rendering, and you opened a can of warms for sure. Nice excavations.

File: packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js [hunk 1]
```diff
@@ -173,13 +173,12 @@ describe('more than one sequence diagram', () => {
 describe('when parsing a sequenceDiagram', function () {
   beforeEach(function () {
     // diagram.db = sequenceDb;
-    // diagram.db.clear();
+    //
```
Comment (nirname) [inline packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js:176]: May be, we do not need those comments at all

**Review by nirname — APPROVED** [review]

File: packages/mermaid/src/Diagram.ts [hunk 2]
```diff
@@ -18,6 +19,7 @@ export class Diagram {
   parser;
   renderer;
   db;
+  private init?: (config: MermaidConfig) => void;
```
Comment (aloisklink) [inline packages/mermaid/src/Diagram.ts:22]: Add an `import { type DiagramDefinition } from './diagram-api/types.js';` to this file, and do:

```suggestion
  parser: DiagramDefinition["parser"];
  renderer: DiagramDefinition["renderer"];
  db: DiagramDefinition["db"];
  private init?: DiagramDefinition["init"];
```

File: packages/mermaid/src/mermaidAPI.ts [hunk 3]
```diff
@@ -108,8 +108,8 @@ export interface RenderResult {
 async function parse(text: string, parseOptions?: ParseOptions): Promise<boolean> {
   addDiagrams();
   try {
-    const diagram = await getDiagramFromText(text);
-    diagram.parse();
+    await getDiagramFromText(text);
+    // diagram.parse();
```
Comment (aloisklink) [inline packages/mermaid/src/mermaidAPI.ts:112]: Please delete instead of turning into a comment, if this is no longer needed.

**Review by aloisklink — COMMENTED** [review]
Wooo, the performance impact sounds pretty nice!

The only potential issue I see is whether or not this is a **BREAKING CHANGE**.

In some of the tests, you've removed lines that used to have `await mermaidAPI.parse(str1);`. Is this something that might break how Mermaid's public API works?

**Review by Yokozuna59 — COMMENTED** [review]
> The only potential issue I see is whether or not this is a **BREAKING CHANGE**.
> 
> In some of the tests, you've removed lines that used to have `await mermaidAPI.parse(str1);`. Is this something that might break how Mermaid's public API works?

I guess it's not a breaking change. When initializing new diagram using `new Diagram()`, it runs the `parse` function async and update the DB, which is exactly what the `mermaidAPI` do when calling `mermaidAPI.parse`.

**Comment by Yokozuna59** [comment]
After looking at these changes, doesn't that mean that there's no use of `text` parameter in `draw` function expect logging what diagram is being rendered with the actual diagram text?

I guess we could add that log call inside the render function with `id`, something like this:

```ts
log.debug(`rendering ${id} diagram\n${text}`);
```

So maybe now we could remove unnecessary parameters in `v11`, see https://github.com/mermaid-js/mermaid/pull/4450#issuecomment-1656153439.

File: packages/mermaid/src/Diagram.ts [hunk 4]
```diff
@@ -61,6 +59,11 @@ export class Diagram {
       throw this.detectError;
     }
     this.db.clear?.();
+    if (this.init) {
+      const config = configApi.getConfig();
+      this.init(config);
+      log.info('Initialized diagram ' + this.type, config);
```
[the remark this answers is not in the document]
Comment (sidharthv96) [reply packages/mermaid/src/Diagram.ts:65]: Removed

File: packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js (hunk 1 above)
Comment (sidharthv96) [reply packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js:176]: Removed

**Comment by sidharthv96** [comment]
> The only potential issue I see is whether or not this is a **BREAKING CHANGE**.
> 
> In some of the tests, you've removed lines that used to have `await mermaidAPI.parse(str1);`. Is this something that might break how Mermaid's public API works?

I don't think it'll have any external changes visible to the mermaidAPI.parse users. 

---

> So maybe now we could remove unnecessary parameters in v11, see https://github.com/mermaid-js/mermaid/pull/4450#issuecomment-1656153439.

Yes, we could! The renderer should not need to know anything regarding the original text. It's only concern should be rendering whatever is in the DB.

**Review by aloisklink — APPROVED** [review]
> I don't think it'll have any external changes visible to the `mermaidAPI.parse` users.

In that case, I'll officially approve this PR :)

---

> > So maybe now we could remove unnecessary parameters in v11, see https://github.com/mermaid-js/mermaid/pull/4450#issuecomment-1656153439.
>
> Yes, we could! The renderer should not need to know anything regarding the original text. It's only concern should be rendering whatever is in the DB.

Should we make a GitHub issue (or maybe a GitHub Discussion is better) for all of the breaking changes we want in a v11 release?

There are some breaking changes I want to make to Mermaid.JS types :)

**Comment by aloisklink** [comment]
Side-note, what label should we give this PR for the Release notes? `Type: Bug / Error` or `Type: Other`?

Maybe we should make a new label called `Type: Performance`, since most JavaScript projects use a different label for performance stuff (e.g. see [`@commitlint/config-conventional`](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional), which uses `perf: xxxx` for commits that affect performance).

**Comment by Yokozuna59** [comment]
> Should we make a GitHub issue (or maybe a GitHub Discussion is better) for all of the breaking changes we want in a v11 release?

I think we should. We are currently just suggesting stuff in random issues and PRs without collecting and filtering them. It would also help us create a changelog for v11.

**Comment by sidharthv96** [comment]
Done: https://github.com/orgs/mermaid-js/discussions/4710
Also updated the `next` branch, which we could use for the breaking chages.

**Review by knsv — APPROVED** [review]
Good Job @sidharthv96 !

**Pull request merged**